fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java

```diff
@@ -773,19 +772,47 @@ private boolean tryLock(boolean mustLock) {
                 }
                 return true;
             } catch (InterruptedException e) {
-                LOG.warn("got exception while getting catalog lock", e);
-                if (mustLock) {
-                    continue;
-                } else {
-                    return lock.isHeldByCurrentThread();
+                LOG.warn("got exception while getting catalog read lock", e);
+                if (!mustLock) {
+                    return false;
                 }
             }
         }
     }
 
-    private void unlock() {
-        if (lock.isHeldByCurrentThread()) {
-            this.lock.unlock();
+    private boolean tryWriteLock(boolean mustLock) {
+        while (true) {
+            try {
+                if (!rwLock.writeLock().tryLock(Config.catalog_try_lock_timeout_ms, TimeUnit.MILLISECONDS)) {
+                    // to see which thread held this lock for long time.
+                    Thread owner = rwLock.getOwner();
+                    if (owner != null) {
+                        LOG.warn("catalog write lock is held by: {}", Util.dumpThread(owner, 50));
+                    }
+
+                    if (mustLock) {
+                        continue;
+                    } else {
+                        return false;
+                    }
+                }
+                return true;
+            } catch (InterruptedException e) {
+                LOG.warn("got exception while getting catalog write lock", e);
+                if (!mustLock) {
+                    return rwLock.isWriteLockedByCurrentThread();
+                }
+            }
+        }
+    }
+
+    private void readUnlock() {
+        this.rwLock.readLock().unlock();
+    }
+
+    private void writeUnlock() {
+        if (this.rwLock.writeLock().isHeldByCurrentThread()) {
```

Comment:
Is this judge wisdom? Lock and unlock must appear in pairs, this should be guaranteed by the caller. I think this judge will hide some problems.